### PR TITLE
[[FEAT]] Replacing PhantomJS with Puppeteer

### LINKED ISSUE #3391: phantomjs is an unmaintained project
Hello Team,

PhantomJS is an unmaintained project and jshint uses phantomjs for its browser testing.

I am curious to know that will you be interested in replacing the phantomjs with some other suitable alternative, as it is better using up to date dependencies.

Please share your thoughts on this and any further suggestion regarding the phantomjs replacement.

### PR DESCRIPTION
**Pull request by ossdev07**
PhantomJS is an unmaintained project replaced Phantomjs with puppeteer.

Fixes #3391

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -5,7 +5,7 @@ before_install:
   # they are only needed to verify correctness in a browser-like environment,
   # they can be removed (and the tests skipped) without effecting coverage.
   - if [[ $(node --version) != v10* ]] ; then
-      node scripts/remove-dev-dependencies.js phantom phantomjs-prebuilt;
+      node scripts/remove-dev-dependencies.js puppeteer;
```
[the remark this answers is not in the document]
Comment (ossdev07) [reply .travis.yml]: Puppeteer is available from Node.js version > 7.0. So, we have to keep this line intact.

File: tests/browser.js [hunk 2]
```diff
@@ -1,57 +1,24 @@
...
-          ph.exit();
-          server.close();
-          if (err) {
-            process.exit(1);
-          }
+const browser = puppeteer.launch({ headless: true, args: ['--no-sandbox', '--disable-setuid-sandbox'] }).then(async browser => {
+        const page = await browser.newPage();
+        createTestServer(port, function(server) {
+        page.on('console', msg => console.log("", msg.text()));
+        page.exposeFunction('callChrome', text=> {
```
[the remark this answers is not in the document]
Comment (ossdev07) [reply tests/browser.js]: Modified and added a commit with changes indentation and linting rules with phantom environment option. Please check and provide feedback if any more modifications required.

**Comment by ossdev07** [comment]
@jugglinmike , Please have a look at the commit for changes requested and provide feedback if any more modifications required.

File: examples/.jshintrc [hunk 3]
```diff
@@ -76,7 +76,7 @@
     "mootools"      : false,    // MooTools
     "node"          : false,    // Node.js
     "nonstandard"   : false,    // Widely adopted globals (escape, unescape, etc)
-    "phantom"       : false,    // PhantomJS
+    "chrome"        : false,    // Chrome
```
Comment (jugglinmike) [inline examples/.jshintrc]: We should keep `phantom`, and we shouldn't add `chome`. Could you revert this change?

File: src/vars.js [hunk 4]
```diff
@@ -533,7 +533,7 @@ exports.browserify = {
 };
 
 exports.phantom = {
-  phantom      : true,
+  phantom       : true,
```
Comment (jugglinmike) [inline src/vars.js]: Could you preserve the original alignment of these properties?

File: tests/browser.js [hunk 5]
```diff
@@ -1,57 +1,33 @@
...
-    "not available. Please note that JSHint does not support development " +
-    "using versions of Node.js which are no longer maintained."
-  );
+  puppeteer = require('puppeteer');
+} catch (err) {	
+  throw new Error(	
+    "Unable to run tests in Chrome because the required dependencies are " +	
+    "not available. Please note that JSHint does not support development " +	
+    "using versions of Node.js which are no longer maintained."	
+  );
```
Comment (jugglinmike) [inline tests/browser.js]: It's hard to see from GitHub.com, but this patch is adding trailing whitespace to some lines in this file. Could you remove the trailing whitespace?

**Comment by pruthvireddy1991** [comment]
@jugglinmike ,

Please have a look at the commit for changes requested and provide feedback if any more modifications required

**Comment by ossdev07** [comment]
@jugglinmike , Please have a look at the commit for changes requested and provide feedback if any more modifications required.

**Comment by jugglinmike** [comment]
Excellent work, @ossdev07! There were a few tiny formatting bugs remaining, but you've already been very patient with the review for this patch, so I thought I'd help out and make the corrections myself:

https://github.com/jshint/jshint/commit/c9c746aad2e4eed66362e0eeaa5c48e72cc46759

In order to preserve the attribution in the project's git history, I merged our work to `master` using a merge commit. One way to visualize that is using the `--graph` option to the `git log` command:

<details>
  <summary><code>$ git log --graph -n 6</code></summary>

    *   commit a3a556e6d1afe4e85b47e5233ef1ea5988a139b6 (HEAD -> master, upstream/master)
    |\  Merge: b7dd7d8f c9c746aa
    | | Author: Mike Pennisi <mike@mikepennisi.com>
    | | Date:   Tue Jan 14 20:07:58 2020 -0500
    | | 
    | |     [[CHORE]] Replace PhantomJS with Puppeteer
    | | 
    | * commit c9c746aad2e4eed66362e0eeaa5c48e72cc46759 (ossdev07-jshint_puppeteer)
    | | Author: Mike Pennisi <mike@mikepennisi.com>
    | | Date:   Tue Jan 14 19:59:41 2020 -0500
    | | 
    | |     Revert a few unintentional formatting changes
    | | 
    | * commit 5463470cc60dbd8a3faed4392493f06113149916
    | | Author: ossdev <ossdev@puresoftware.com>
    | | Date:   Mon Jan 13 12:51:39 2020 +0530
    | | 
    | |     Updated linting rules
    | | 
    | * commit fa8ad167adec95bb8931f31e2cb937b6817f90cc
    | | Author: ossdev07 <ossdev@puresoftware.com>
    | | Date:   Mon Dec 16 15:05:09 2019 +0530
    | | 
    | |     Added linting rules for puppeteer
    | | 
    | * commit 51963a34ba6593cdbfe208fa4569879a182bafe2
    | | Author: ossdev07 <ossdev@puresoftware.com>
    | | Date:   Fri Dec 13 12:33:51 2019 +0530
    | | 
    | |     [[FEAT]] Replacing PhantomJS with Puppeteer
    | |     
    | |     Signed-off-by: ossdev07 <ossdev@puresoftware.com>
    | |   
    * |   commit b7dd7d8f93244abb874fdb616c7a5725209e36b0 (tag: 2.11.0)
    |\ \  Merge: 69767ed2 f7d3a462
    | |/  Author: jugglinmike <mike@mikepennisi.com>
    |/|   Date:   Mon Jan 13 14:35:50 2020 -0500
    | |   
    | |       Merge pull request #3432 from jshint/v2.11.0
    | |       
    | |       Version 2.11.0

</details>

If you expand the terminal output above, you'll see how your commits and my little nitpicks both occur in a separate branch of development, and a final "merge" commit brings them all into the project's `master` branch.

Thanks to you, we no longer have to worry about bugs in the unmaintained PhatomJS project. That's great! You first reported this issue almost nine months ago--your commitment is impressive :)

Thanks again
